### PR TITLE
Feat/sqlc integration

### LINKED ISSUE #17: add sqlc integration

### LINKED ISSUE #18: change db schema to a revised one
```sql
CREATE TABLE decryption_keys_message (
    slot                  BIGINT PRIMARY KEY,
    instance_id     BIGINT NOT NULL,
    eon                  BIGINT NOT NULL,
    tx_pointer       BIGINT NOT NULL
);
```
```sql
CREATE TABLE decryption_key (
    eon                                   BIGINT,
    identity_preimage           BYTEA,
    key                                    BYTEA NOT NULL,
    PRIMARY KEY                 (eon, identity_preimage)
);
```
```sql
CREATE TABLE decryption_keys_message_decryption_key (
    decryption_keys_message_slot              BIGINT,
    key_index                                                   BIGINT,
    decryption_key_eon                                 BIGINT,
    decryption_key_identity_preimage         BYTEA,
    PRIMARY KEY (decryption_keys_message_slot, decryption_key_eon, decryption_key_identity_preimage, key_index),
    FOREIGN KEY (decryption_keys_message_slot) REFERENCES decryption_keys_message (slot),
    FOREIGN KEY (decryption_key_eon, decryption_key_identity_preimage) REFERENCES decryption_key (eon, identity_preimage)
);
```
```sql
CREATE TABLE transaction_submitted_events (
    event_block_hash                       BYTEA,
    event_block_number                  BIGINT,
    event_tx_index                            BIGINT,
    event_log_index                         BIGINT,
    eon                                               BIGINT NOT NULL,
    tx_index                                      BIGINT NOT NULL,
    identity_prefix                            BYTEA  NOT NULL,
    sender                                         BYTEA  NOT NULL,
    encrypted_transaction              BYTEA  NOT NULL,
    PRIMARY KEY (event_block_hash, event_block_number, event_tx_index, event_log_index)
);
```

(had a call with fred about this and I think we're mostly aligned, but having a code example makes it easier to talk about)

to get all decryption keys for a slot this query should work:

```sql
SELECT
    whatever
FROM decryption_keys_message_decryption_key dkmdk
LEFT JOIN decryption_keys_message dkm ON dkmdk.decryption_keys_message_id = dkm.id
LEFT JOIN decryption_key dk ON dkmdk.identity_preimage = dk.identity_preimage
WHERE dkm.slot = $1;
```
and then you can take the corresponding encrypted txs with

```sql
SELECT * FROM encrypted_tx
WHERE eon = $1 AND index >= $2 AND index < $2 + $3
```

### PR DESCRIPTION
**Pull request by faheelsattar**
closes #17 
closes #18

### CODE REVIEW

**Comment by faheelsattar** [comment]
need to update in memory tx mapper with the updated schema as well to reflect new data added

File: internal/data/metrics.sql.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
+// Code generated by sqlc. DO NOT EDIT.
```
Comment (fredo) [inline internal/data/metrics.sql.go:1]: nitpick: general question: is it a convention to end generated files with `.gen.go`?

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    depends_on:
+      - db
+      - prometheus
+    environment:
+      - DB_HOST=${DB_HOST}
+      - DB_PORT=${DB_PORT}
+      - DB_USER=${DB_USER}
+      - DB_PASSWORD=${DB_PASSWORD}
+      - DB_NAME=${DB_NAME}
+      - DB_SSL_MODE=${DB_SSL_MODE}
```
Comment (ulope) [inline docker-compose.yml:18]: AFAICS we don't need those to be configurable (with the exception of DB_PASSWORD maybe) from the "outside".

**Review by ulope — COMMENTED** [review]
Addes some comments below.

I can't say much on the actual implementation other the devops parts since I don't know the history of that part all that well.

Fred said he would look into it.

File: cmd/cli/cli.go [hunk 3]
```diff
@@ -26,7 +32,36 @@ func Cmd() *cobra.Command {
...
+				bootstrapP2PAddresses[i] = address.MustP2PAddress(addr)
+			}
+			config.P2P.CustomBootstrapAddresses = bootstrapP2PAddresses
+			config.P2P.ListenAddresses = []*address.P2PAddress{
+				address.MustP2PAddress("/ip4/0.0.0.0/tcp/0"),
+				address.MustP2PAddress("/ip4/0.0.0.0/udp/0/quic-v1"),
+				address.MustP2PAddress("/ip4/0.0.0.0/udp/0/quic-v1/webtransport"),
+				address.MustP2PAddress("/ip6/::/tcp/0"),
+				address.MustP2PAddress("/ip6/::/udp/0/quic-v1"),
+				address.MustP2PAddress("/ip6/::/udp/0/quic-v1/webtransport"),
```
[the remark this answers is not in the document]
Comment (faheelsattar) [reply cmd/cli/cli.go]: fixed

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+version: '3.8'
```
[the remark this answers is not in the document]
Comment (faheelsattar) [reply docker-compose.yml]: removed

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+services:
+  app:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    container_name: gnosh-observer
+    depends_on:
+      - db
+      - prometheus
```
[the remark this answers is not in the document]
Comment (faheelsattar) [reply docker-compose.yml]: removed

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      - DB_PORT=${DB_PORT}
+      - DB_USER=${DB_USER}
+      - DB_PASSWORD=${DB_PASSWORD}
+      - DB_NAME=${DB_NAME}
+      - DB_SSL_MODE=${DB_SSL_MODE}
+    ports:
+      - "8080:8080"
+    networks:
+      - app-network
+    command: ["./observer", "start", "--rpc-url", "wss://rpc.chiadochain.net/wss", "--p2pkey", "CAESQMl3XPvSPMdLdbVg0M3/5ZenGuk5+Ve0diP0i3F9WH76Xw3ax7KRmBm8CjOcv8FpfINtlW9NrIpsWjeliuGybew="]
```
[the remark this answers is not in the document]
Comment (faheelsattar) [reply docker-compose.yml]: got it

File: internal/data/metrics.sql.go (hunk 1 above)
Comment (faheelsattar) [reply internal/data/metrics.sql.go:1]: don't think if its a convention, but makes it easier to understand that this file was auto generated, and don't edit it directly.

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    depends_on:
+      - db
+    environment:
+      - DB_HOST=db
+      - DB_PORT=5432
+      - DB_USER=postgres
+      - DB_PASSWORD=${DB_PASSWORD}
+      - DB_NAME=shutter_metrics
+      - DB_SSL_MODE=disable
+      - P2P_BOOTSTRAP_ADDRESSES=${P2P_BOOTSTRAP_ADDRESSES}
```
Comment (ylembachar) [inline docker-compose.yml:16]: would be great to have a template env file to copy with all the env variables and potential default values

File: internal/watcher/p2p_msgs.go [hunk 8]
```diff
@@ -52,36 +56,37 @@ func NewP2PMsgsWatcherWatcher(
...
-func (dkw *P2PMsgsWatcher) MessagePrototypes() []p2pmsg.Message {
+func (pmw *P2PMsgsWatcher) MessagePrototypes() []p2pmsg.Message {
 	return []p2pmsg.Message{
 		&p2pmsg.DecryptionKeys{},
 		&p2pmsg.DecryptionKeyShares{},
 	}
 }
 
-func (dkw *P2PMsgsWatcher) ValidateMessage(_ context.Context, msgUntyped p2pmsg.Message) (pubsub.ValidationResult, error) {
+func (pmw *P2PMsgsWatcher) ValidateMessage(_ context.Context, msgUntyped p2pmsg.Message) (pubsub.ValidationResult, error) {
```
Comment (fredo) [inline internal/watcher/p2p_msgs.go:82]: I think we should validate the messages a bit more, otherwise fake messages can easily mess with the metrics. But this can be left for another PR

File: cmd/cli/cli.go [hunk 9]
```diff
@@ -36,7 +72,7 @@ func Cmd() *cobra.Command {
 		"rpc-url",
 		"",
 		"wss://rpc.chiadochain.net/wss",
-		"gnosis testnet rpc url",
+		"gnosis websocket rpc url",
```
[the remark this answers is not in the document]
Comment (faheelsattar) [reply cmd/cli/cli.go:73]: required the cli args for rpc and contract address

File: internal/watcher/p2p_msgs.go (hunk 8 above)
Comment (faheelsattar) [reply internal/watcher/p2p_msgs.go:82]: alright we can do that
